### ISSUE TITLE
Memory corruption observed when using Moka v0.9.6

### ISSUE THREAD

**Issue by tatsuya6502** [moka-rs/moka#281]
Split from #279.

> Note: I started looking into Moka because recently we tried using it in https://github.com/MaterializeInc/materialize/pull/19614, but had to revert it because we were observing memory corruption.

- Moka v0.9.6
- No stack traces and core dumps are available.
- The target platform (CPU arch and OS) ~~are yet unknown.~~ => Linux x86_64
- `moka::sync::Cache` are used in async contexts.
- Multiple (many?)`Cache` instances. (An instance per blob URL. Do not know how many)
- Weigher and eviction listener (immediate mode) are configured.

I am trying to reproduce the issue using mokabench but no luck: https://github.com/moka-rs/moka/issues/279#issuecomment-1596820321

CC: @ParkMyCar

**Comment by tatsuya6502** [comment]
Details of the issue: https://github.com/MaterializeInc/materialize/issues/19746

It seems easy to reproduce if we can run the integration test called "long Zippy with user tables".

> This happened twice in a row in long Zippy w/ user tables test: https://buildkite.com/materialize/release-qualification/builds/197#018883bf-5086-4a43-a8b4-ede1ef192de5

**Comment by tatsuya6502** [comment]
https://github.com/MaterializeInc/materialize/blob/bfbebe04a7d181f1bac91f01785309a52cf2de34/ci/release-qualification/pipeline.yml#L99-L108

```yml
  - id: zippy-user-tables-large
    label: "Long Zippy w/ user tables"
    timeout_in_minutes: 2880
    agents:
      queue: linux-x86_64-large
    artifact_paths: junit_*.xml
    plugins:
      - ./ci/plugins/mzcompose:
          composition: zippy
          args: [--scenario=UserTablesLarge, --actions=200000]
```

- CI Overview: https://github.com/MaterializeInc/materialize/blob/main/ci/README.md

**Comment by tatsuya6502** [comment]
There are almost 300 commits between when moka was added and when removed:
https://github.com/MaterializeInc/materialize/compare/1bcd1c4b89...07386e427f

**Comment by tatsuya6502** [comment]
> There are almost 300 commits between when moka was added and when removed:
> https://github.com/MaterializeInc/materialize/compare/1bcd1c4b89...07386e427f
> 
> I need to know when (which commit) last failed.

The following commits were mentioned in https://github.com/MaterializeInc/materialize/issues/19746.

- https://github.com/MaterializeInc/materialize/commit/9ab5f833b919f8a5200b3918eca0847f40cc3337 (Jun 4, 2023)
    - "Long Zippy w/ user tables" failed.
- https://github.com/MaterializeInc/materialize/commit/3deb5a4739c4e67d677068322b52ea75ddf8252e (Jun 13, 2023)
    - "testdrive" failed.
    - [BuildKite log](https://buildkite.com/materialize/nightlies/builds/2585#0188b1f3-b522-4219-aad7-bc7b0a5783f2)

**Comment by tatsuya6502** [comment]
> - https://github.com/MaterializeInc/materialize/commit/9ab5f833b919f8a5200b3918eca0847f40cc3337 (Jun 4, 2023)
>     - "Long Zippy w/ user tables" failed.

Ran the test (with 5,000 actions) several times on my home PC, but I have not reproduced the issue.

> **Environment:**
>
> - Docker (x86_64)
> - Intel Core i7-12700F
>     - 8 × performance cores × 2 hyper threads
>     - 4 × efficient cores

My PC (Windows 11 + WSL2) does not have enough RAM to run longer Zippy test. It has only 16GB of RAM and Docker containers like Kafka sometimes get killed while running test. I will expand the RAM to 32GB.

I hope running longer test would help to reproduce the issue. But another aspect could be the number of available cores/hyper threads. My PC has 16 cores and 20 threads, but their EC2 instance only has 8 vCPUs:

- c5.2xlarge instances, which have 8 vCPUs (i.e., hyperthreads) and 16GB of memory.

I will probably adjust the number of hyper threads available for Docker:

- https://docs.docker.com/desktop/settings/windows/
    - > In WSL 2 mode, you can configure limits on the memory, CPU, and swap size allocated to the [WSL 2 utility VM](https://docs.microsoft.com/en-us/windows/wsl/wsl-config#configure-global-options-with-wslconfig).

Also, I will add some tracing logs to both Moka and Materialize to get better understandings on the access pattern.

**Comment by 0xdeafbeef** [comment]
@tatsuya6502 I have 32 threads /  128 ram and can leave test running for night if this can help

**Comment by tatsuya6502** [comment]
@0xdeafbeef Thanks for the offer. Wow, you have a great machine!

I think I am going to be okay. I already ordered the RAM modules and am waiting for them to arrive. I think 32GB should be enough for now.

**Comment by tatsuya6502** [comment]
FYI, I got the RAM modules for my PC, and confirmed it can run the test with at least 50,000 actions. I allocated 24GiB of memory and only 8 logical CPU cores to WSL2.

The reason I did not try 200,000 actions is that it will take long time (Estimated ~30 hours). 50,000 actions took 7 hours and 14 minutes. The memory utilization was fine so I had feeling that it could run 200,000 without issue.

Also, before I got the RAM module, I was also able to run 80,000 actions. I found that reducing the number of logical CPU cores for WSL2/Docker will reduce the memory utilization of the test. So I allocated only 8 cores (instead of 20) and 10GiB of memory (instead of 8GiB) to WSL2 and it was able to run 80,000 actions.

**Comment by tatsuya6502** [comment]
I talked to a Materialize engineer. They will try to reproduce the issue using an EC2 instance and give me the full set of instructions.

https://github.com/MaterializeInc/materialize/discussions/20136#discussioncomment-6274272

> Hopefully someone from the QA team has time to attempt the repro early next week and give you the full set of instructions.

**Comment by tatsuya6502** [comment]
With some helps from Materialize engineers, I made some progress. I reproduced the segmentation fault in Moka v0.9.6 in Materialize, and got a core dump to analyze.

NOTES: I had to use an Amazon EC2 `c6a.12xlarge` instance, which has 48 × vCPUs. This is because using my home PC (20 × HyperThreads) can take few days to reproduce with the current test ([details](https://github.com/MaterializeInc/materialize/discussions/20136#discussioncomment-6279920)). Also it was a bit tricky to get core dump in a Docker container, so it took some time to me to figure out. ([An modified Materialize test harness for getting core dumps](https://gitlab.com/moka-labs/moka-gh279-materialize-memory-corruption/materialize/-/commit/ed122687fb3b4755c2d2a7443c8e706ecab14d98))

Here is the back trace of the crashed thread:

![moka-segfault-2023-07-01](https://github.com/moka-rs/moka/assets/159403/2a662ceb-08e8-4409-b5af-03ae3a6bff93)

It crashed in the `Deque::move_to_back` method ([src/common/deque.rs#L202](https://github.com/moka-rs/moka/blob/v0.9.6/src/common/deque.rs#L202)), which was called by the `base_cache::Inner::handle_upsert` method ([src/sync_base/base_cache.rs#L1414](https://github.com/moka-rs/moka/blob/v0.9.6/src/sync_base/base_cache.rs#L1414)).

Materialize uses Moka v0.9.6, and this problem may have been fixed by [#259](https://github.com/moka-rs/moka/pull/259) in v0.11.0.

- #259

I will back-port the fix to v0.9.x, and run the same test to see if the problem goes away.

### Memo

If this is related to [#259](https://github.com/moka-rs/moka/pull/259), a reproducing step could be the followings, but the end result depends largely on how the compiler optimizes reading an immutable reference `&` to a LRU double-ended queue (deque).

1. Create a cache with weighted capacity 6.
2. Stop the periodic `sync`.
3. Get not-existing key `X` once. (popularity = 1)
4. Get not-existing key `Y` once. (popularity = 1)
5. Insert key `X` with weight 2.
6. Insert key `Y` with weight 2.
7. Insert key `Z` with weight 2.
8. Call `sync`:
    - All `X`, `Y` and `Z` are admitted.
    - Now the LRU deque has: (front) `X` →  `Y` → `Z` (back)
9. Get not-existing key `V` three times. (popularity = 3)
10. Insert `V` with weight 3.
11. Invalidate `Y`. 
    - This removes `Y` from the concurrent hash table (cht), but removing it’s node from the deque will be delayed to a future `sync`.
12. Call `sync`. The followings will happen and the last one can cause a segmentation fault:
    - `X` and `Y` are selected as the victims (entries to evict), and `V` is marked as admitted.
        - weights: `X` + `Y` = 4, `V` = 3
        - popularity: `X` + `Y` = 2, `V` = 3
    - `X` is removed from the cht, and its deque node is unlinked and dropped.
        - Unlinking `X` node changes `Y` node's `prev` pointer from `Some(NonNull(X))` to `None` .
    - Because of step 11, `Y` is no longer in the cht, so `Y`'s deque node is added to the `skipped_nodes` vector.
    - `V`'s deque node is created and pushed to the back of the deque.
    - The nodes in the `skipped_node` is processed: `Y` node will be moved to the back of the deque.
        - Deque's `move_to_back` method checks `Y`'s `prev` pointer, which is `None` now.
        - However, because of the bug (fixed by [#259](https://github.com/moka-rs/moka/pull/259) for v0.11.0), optimized machine instructions may use an stale value of the `prev` pointer in a CPU register (pointing to `X`), instead of reloading the actual value `None` from the memory.
        - `X` was already dropped, so trying to access `X` (in the following code: `*prev.as_ptr()`) will cause the segmentation fault.

[src/common/deque.rs#L202](https://github.com/moka-rs/moka/blob/v0.9.6/src/common/deque.rs#L202)
```rust
188: pub(crate) unsafe fn move_to_back(&mut self, mut node: NonNull<DeqNode<T>>) {
        ...
201:    match node.prev {
202:        Some(prev) if node.next.is_some() => (*prev.as_ptr()).next = node.next,
```

**Comment by tatsuya6502** [comment]
> 11. Invalidate `Y`. 
>     - This removes `Y` from the concurrent hash table (cht), but removing it’s node from the deque will be delayed to a future `sync`.
> 12. Call `sync`. The followings will happen and the last one can cause a segmentation fault:
>     - `X` and `Y` are selected as the victims (entries to evict), and `V` is marked as admitted.
>         - weights: `X` + `Y` = 4, `V` = 3
>         - popularity: `X` + `Y` = 2, `V` = 3
>     - `X` is removed from the cht, and its deque node is unlinked and dropped.
>         - Unlinking `X` node changes `Y` node's `prev` pointer from `Some(NonNull(X))` to `None` .
>     - Because of step 11, `Y` is no longer in the cht, so `Y`'s deque node is added to the `skipped_nodes` vector.

I reviewed code and realized that above will not reproduce the problem.

- Step 11 (invalidate `Y`) needs to be done in a smaller time window.
- It has to be done _after_ the first bullet of 12 (`X` and `Y` are selected as ...) and _before_ the third bullet of 12 (Because of step 11, `Y` is no longer in the cht, ...).

**Comment by tatsuya6502** [comment]
I will backport #259 to Moka v0.10.x as well. I will keep this issue open until I am done with it.

**Comment by tatsuya6502** [comment]
Closing as completed.

**Moka v0.9.x**

- Backported [#259](https://github.com/moka-rs/moka/pull/259) to Moka v0.9.8 via #283.
- Published [Moka v0.9.8](https://crates.io/crates/moka/0.9.8) to crates.io.

**Moka v0.10.x**

- Backported [#259](https://github.com/moka-rs/moka/pull/259) to Moka v0.10.3 via #286.
- Published [Moka v0.10.3](https://crates.io/crates/moka/0.10.3) to crates.io.
